libs/ngxtension/create-injection-token/src/create-injection-token.ts

```diff
@@ -0,0 +1,153 @@
+import {
+	Host,
+	InjectionToken,
+	Optional,
+	Self,
+	SkipSelf,
+	inject,
+	type FactoryProvider,
+	type InjectOptions,
+	type Provider,
+	type Type,
+} from '@angular/core';
+
+type CreateInjectionTokenDep<TTokenType> =
+	| Type<TTokenType>
+	// NOTE: we don't have an AbstractType
+	| (abstract new (...args: any[]) => TTokenType)
+	| InjectionToken<TTokenType>;
+
+type CreateInjectionTokenDeps<
+	TFactory extends (...args: any[]) => any,
+	TFactoryDeps extends Parameters<TFactory> = Parameters<TFactory>
+> = {
+	[Index in keyof TFactoryDeps]:
+		| CreateInjectionTokenDep<TFactoryDeps[Index]>
+		| [
+				...modifiers: Array<Optional | Self | SkipSelf | Host>,
+				token: CreateInjectionTokenDep<TFactoryDeps[Index]>
+		  ];
+} & { length: TFactoryDeps['length'] };
+
+export type CreateInjectionTokenOptions<
+	TFactory extends (...args: any[]) => any,
+	TFactoryDeps extends Parameters<TFactory> = Parameters<TFactory>
+> =
+	// this means TFunction has no arguments
+	(TFactoryDeps[0] extends undefined
+		? {
+				isRoot: boolean;
+				deps?: never;
+		  }
+		: {
+				isRoot?: boolean;
+				deps: CreateInjectionTokenDeps<TFactory, TFactoryDeps>;
+		  }) & {
+		token?: InjectionToken<ReturnType<TFactory>>;
+		extraProviders?: Provider;
+	};
+
+type InjectFn<
+	TFactory extends (...args: any[]) => any,
+	TFactoryReturn extends ReturnType<TFactory> = ReturnType<TFactory>
+> = {
+	(): TFactoryReturn;
+	(injectOptions: InjectOptions & { optional?: false }): TFactoryReturn;
+	(injectOptions: InjectOptions): TFactoryReturn | null;
+};
+
+export type CreateInjectionTokenReturn<
+	TFactory extends (...args: any[]) => any,
+	TFactoryReturn extends ReturnType<TFactory> = ReturnType<TFactory>
+> = [
+	InjectFn<TFactory, TFactoryReturn>,
+	(value?: TFactoryReturn) => Provider,
+	InjectionToken<TFactoryReturn>
+];
+
+function createInjectFn<TValue>(token: InjectionToken<TValue>) {
+	return (injectOptions?: InjectOptions) =>
+		inject(token, injectOptions as InjectOptions);
+}
+
+function createProvideFn<
+	TValue,
+	TFactory extends (...args: any[]) => any = (...args: any[]) => TValue,
+	TFactoryDeps extends Parameters<TFactory> = Parameters<TFactory>
+>(
+	token: InjectionToken<TValue>,
+	factory: (...args: any[]) => TValue,
+	deps?: CreateInjectionTokenDeps<TFactory, TFactoryDeps>,
+	extraProviders?: Provider
+) {
+	return (value?: TValue) => {
+		let provider: Provider;
+		if (value) {
+			provider = { provide: token, useValue: value };
+		} else {
+			provider = {
+				provide: token,
+				useFactory: factory,
+				deps: (deps ?? []) as FactoryProvider['deps'],
+			};
+		}
+
+		return extraProviders ? [extraProviders, provider] : provider;
```

Comment:
Do we care about cases where we want to create EnvironmentProviders ? 

Maybe another param? 
Would it be better if we add an object for options and not leave them as separate param ?

Reply:
what's the difference between Provider and EnvironmentProvider in terms of consuming it?